app/src/studio/core.c

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2024 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <zmk/studio/core.h>
+
+ZMK_EVENT_IMPL(zmk_studio_core_lock_state_changed);
+ZMK_EVENT_IMPL(zmk_studio_core_unlock_requested);
+
+static enum zmk_studio_core_lock_state state = IS_ENABLED(CONFIG_ZMK_STUDIO_LOCKING)
+                                                   ? ZMK_STUDIO_CORE_LOCK_STATE_LOCKED
+                                                   : ZMK_STUDIO_CORE_LOCK_STATE_UNLOCKED;
+
+enum zmk_studio_core_lock_state zmk_studio_core_get_lock_state(void) { return state; }
+
+static void set_state(enum zmk_studio_core_lock_state new_state) {
+    state = new_state;
+
+    raise_zmk_studio_core_lock_state_changed(
```

Comment:
Maybe worth a check to see if the `state != new_state` before raising the event?